### PR TITLE
Taint analysis bug fixes and improvements

### PR DESCRIPTION
**Pull request by topolik**
Hi @h3xstream  and @formanek,

here are the changes from https://github.com/topolik/find-sec-bugs/commit/c173a9766b73438979232f07ee4f4609829aae1a#commitcomment-18322174

There are no Liferay signatures yet.

@formanek: I kept only what we agreed on in the email conversions, that means:
* Fixed the type cast bug
* Fixed a new bug - configured method summaries were ignored when return type was one of the safe objects
* Fixed missing/inconsistent LDAP_INJECTION_SAFE tag
* Added support for multiple taint analysis config files
* Added support for custom (multiple) injection sinks config files
* Added support for config files to be loaded from file system and class path
* Added support for config files to be configured also using system environment properties (useful for IDE plugins and CI (using gradle plugin) where system properties cannot be configured)
* Added signatures for JSR-286 portlet spec
* Moved SAFE_OBJECTS into config files as you mentioned to be better solution for custom cast types
* Renamed classes for Taint config files / summaries + some code cleaning and refactoring
* Added one context-sensitive method summary to fine-tune false positives

I know there are many commits, I tried to keep the granularity at readable level. But if you wish I can send several pull requests rebased one on top of the other (once I return from holiday :smile:  )

I tried to run the built-in profiler to see results on about 880 classes. I see no noticeable performance regression, but please feel free to retest. 

With the changes (only > 1000ms):
```
23:26:32.184 [DEBUG] [system.err] 23:26:32.184 [ERROR] [system.err] PROFILE REPORT
23:26:32.185 [DEBUG] [system.err] 23:26:32.185 [ERROR] [system.err]    msecs    #calls usecs/call Class
23:26:32.185 [DEBUG] [system.err] 23:26:32.185 [ERROR] [system.err]     4898      1086      4510 TaintDataflowEngine
23:26:32.185 [DEBUG] [system.err] 23:26:32.185 [ERROR] [system.err]     2114      1096      1929 ValueNumberDataflowFactory
23:26:32.185 [DEBUG] [system.err] 23:26:32.185 [ERROR] [system.err]     1619      1086      1491 IsNullValueDataflowFactory
23:26:32.186 [DEBUG] [system.err] 23:26:32.186 [ERROR] [system.err]     1580      1084      1458 UnconditionalValueDerefDataflowFactory
23:26:32.186 [DEBUG] [system.err] 23:26:32.186 [ERROR] [system.err]     1224      1081      1133 NullDerefAndRedundantComparisonFinder
23:26:32.186 [DEBUG] [system.err] 23:26:32.186 [ERROR] [system.err]     1071       112      9565 FindBadCast2
23:26:32.186 [DEBUG] [system.err] 23:26:32.186 [ERROR] [system.err]     1011      1092       926 TypeDataflowFactory


23:27:52.528 [DEBUG] [system.err] 23:27:52.528 [ERROR] [system.err] PROFILE REPORT
23:27:52.528 [DEBUG] [system.err] 23:27:52.528 [ERROR] [system.err]    msecs    #calls usecs/call Class
23:27:52.528 [DEBUG] [system.err] 23:27:52.528 [ERROR] [system.err]     4872      1086      4486 TaintDataflowEngine
23:27:52.529 [DEBUG] [system.err] 23:27:52.528 [ERROR] [system.err]     2388      1096      2179 ValueNumberDataflowFactory
23:27:52.529 [DEBUG] [system.err] 23:27:52.529 [ERROR] [system.err]     2053      1086      1890 IsNullValueDataflowFactory
23:27:52.529 [DEBUG] [system.err] 23:27:52.529 [ERROR] [system.err]     1706      1084      1574 UnconditionalValueDerefDataflowFactory
23:27:52.529 [DEBUG] [system.err] 23:27:52.529 [ERROR] [system.err]     1640       865      1897 SpecialTypeAnalysis
23:27:52.529 [DEBUG] [system.err] 23:27:52.529 [ERROR] [system.err]     1461      1081      1352 NullDerefAndRedundantComparisonFinder
23:27:52.530 [DEBUG] [system.err] 23:27:52.530 [ERROR] [system.err]     1315       112     11741 FindBadCast2
23:27:52.530 [DEBUG] [system.err] 23:27:52.530 [ERROR] [system.err]     1214       112     10847 FindDeadLocalStores
23:27:52.530 [DEBUG] [system.err] 23:27:52.530 [ERROR] [system.err]     1125      1092      1030 TypeDataflowFactory


23:34:44.566 [DEBUG] [system.err] 23:34:44.565 [ERROR] [system.err] PROFILE REPORT
23:34:44.566 [DEBUG] [system.err] 23:34:44.566 [ERROR] [system.err]    msecs    #calls usecs/call Class
23:34:44.567 [DEBUG] [system.err] 23:34:44.567 [ERROR] [system.err]     5715      1086      5263 TaintDataflowEngine
23:34:44.567 [DEBUG] [system.err] 23:34:44.567 [ERROR] [system.err]     2431      1096      2218 ValueNumberDataflowFactory
23:34:44.568 [DEBUG] [system.err] 23:34:44.567 [ERROR] [system.err]     2012      1086      1853 IsNullValueDataflowFactory
23:34:44.568 [DEBUG] [system.err] 23:34:44.568 [ERROR] [system.err]     1888      1084      1741 UnconditionalValueDerefDataflowFactory
23:34:44.568 [DEBUG] [system.err] 23:34:44.568 [ERROR] [system.err]     1495      1081      1383 NullDerefAndRedundantComparisonFinder
23:34:44.568 [DEBUG] [system.err] 23:34:44.568 [ERROR] [system.err]     1145      1092      1048 TypeDataflowFactory
23:34:44.568 [DEBUG] [system.err] 23:34:44.568 [ERROR] [system.err]     1096       112      9791 FindBadCast2

23:35:55.911 [DEBUG] [system.err] 23:35:55.911 [ERROR] [system.err] PROFILE REPORT
23:35:55.912 [DEBUG] [system.err] 23:35:55.912 [ERROR] [system.err]    msecs    #calls usecs/call Class
23:35:55.912 [DEBUG] [system.err] 23:35:55.912 [ERROR] [system.err]     4896      1086      4508 TaintDataflowEngine
23:35:55.912 [DEBUG] [system.err] 23:35:55.912 [ERROR] [system.err]     2631      1096      2401 ValueNumberDataflowFactory
23:35:55.912 [DEBUG] [system.err] 23:35:55.912 [ERROR] [system.err]     1719      1084      1586 UnconditionalValueDerefDataflowFactory
23:35:55.913 [DEBUG] [system.err] 23:35:55.912 [ERROR] [system.err]     1582      1086      1456 IsNullValueDataflowFactory
23:35:55.913 [DEBUG] [system.err] 23:35:55.913 [ERROR] [system.err]     1577      1081      1459 NullDerefAndRedundantComparisonFinder
```

When I tried the very same on master branch:
```
23:40:23.454 [DEBUG] [system.err] 23:40:23.454 [ERROR] [system.err] PROFILE REPORT
23:40:23.455 [DEBUG] [system.err] 23:40:23.454 [ERROR] [system.err]    msecs    #calls usecs/call Class
23:40:23.455 [DEBUG] [system.err] 23:40:23.455 [ERROR] [system.err]     4979      1086      4585 TaintDataflowEngine
23:40:23.455 [DEBUG] [system.err] 23:40:23.455 [ERROR] [system.err]     2444      1081      2261 NullDerefAndRedundantComparisonFinder
23:40:23.455 [DEBUG] [system.err] 23:40:23.455 [ERROR] [system.err]     2345      1096      2139 ValueNumberDataflowFactory
23:40:23.456 [DEBUG] [system.err] 23:40:23.456 [ERROR] [system.err]     1812      1086      1669 IsNullValueDataflowFactory
23:40:23.456 [DEBUG] [system.err] 23:40:23.456 [ERROR] [system.err]     1778      1084      1640 UnconditionalValueDerefDataflowFactory
23:40:23.456 [DEBUG] [system.err] 23:40:23.456 [ERROR] [system.err]     1223       112     10925 FindBadCast2
23:40:23.457 [DEBUG] [system.err] 23:40:23.457 [ERROR] [system.err]     1177      1092      1077 TypeDataflowFactory


23:41:25.689 [DEBUG] [system.err] 23:41:25.689 [ERROR] [system.err] PROFILE REPORT
23:41:25.689 [DEBUG] [system.err] 23:41:25.689 [ERROR] [system.err]    msecs    #calls usecs/call Class
23:41:25.689 [DEBUG] [system.err] 23:41:25.689 [ERROR] [system.err]     5166      1086      4757 TaintDataflowEngine
23:41:25.690 [DEBUG] [system.err] 23:41:25.689 [ERROR] [system.err]     2142      1096      1955 ValueNumberDataflowFactory
23:41:25.690 [DEBUG] [system.err] 23:41:25.690 [ERROR] [system.err]     1953      1081      1807 NullDerefAndRedundantComparisonFinder
23:41:25.690 [DEBUG] [system.err] 23:41:25.690 [ERROR] [system.err]     1589      1086      1463 IsNullValueDataflowFactory
23:41:25.691 [DEBUG] [system.err] 23:41:25.691 [ERROR] [system.err]     1555      1084      1434 UnconditionalValueDerefDataflowFactory
23:41:25.691 [DEBUG] [system.err] 23:41:25.691 [ERROR] [system.err]     1094      1092      1001 TypeDataflowFactory
23:41:25.691 [DEBUG] [system.err] 23:41:25.691 [ERROR] [system.err]     1080       112      9649 FindBadCast2

23:42:49.762 [DEBUG] [system.err] 23:42:49.762 [ERROR] [system.err] PROFILE REPORT
23:42:49.762 [DEBUG] [system.err] 23:42:49.762 [ERROR] [system.err]    msecs    #calls usecs/call Class
23:42:49.763 [DEBUG] [system.err] 23:42:49.763 [ERROR] [system.err]     5277      1086      4859 TaintDataflowEngine
23:42:49.763 [DEBUG] [system.err] 23:42:49.763 [ERROR] [system.err]     2133      1096      1946 ValueNumberDataflowFactory
23:42:49.763 [DEBUG] [system.err] 23:42:49.763 [ERROR] [system.err]     1764      1084      1627 UnconditionalValueDerefDataflowFactory
23:42:49.764 [DEBUG] [system.err] 23:42:49.764 [ERROR] [system.err]     1690      1086      1556 IsNullValueDataflowFactory
23:42:49.764 [DEBUG] [system.err] 23:42:49.764 [ERROR] [system.err]     1282      1081      1186 NullDerefAndRedundantComparisonFinder
23:42:49.764 [DEBUG] [system.err] 23:42:49.764 [ERROR] [system.err]     1078       112      9631 FindBadCast2
```

Thanks!

### CODE REVIEW

**Comment by h3xstream** [comment]
Cool. I will do a code review today.

I will git clone your branch and review the aggregate changes in IntelliJ. The rebase is not needed.

File: plugin/src/main/java/com/h3xstream/findsecbugs/injection/BasicInjectionDetector.java [hunk 1]
```diff
@@ -79,6 +89,63 @@ public void receiveInjectionPoint(String fullMethodName, InjectionPoint injectio
...
+     *     <li>-Dfindsecbugs.injection.customconfigfile.SqlInjectionDetector="/tmp/sql-custom.txt|SQL_INJECTION_HIBERNATE:/tmp/sql2-custom.txt|SQL_INJECTION_HIBERNATE"</li>
+     *     <li>-Dfindsecbugs.injection.customconfigfile.ScriptInjectionDetector="/tmp/script-engine-custom.txt|SCRIPT_ENGINE_INJECTION:/tmp/el-custom.txt|EL_INJECTION"</li>
+     * </ul>
+     * Example for Windows:<ul>
+     *     <li>-Dfindsecbugs.injection.customconfigfile.SqlInjectionDetector="C:\Temp\sql-custom.txt|SQL_INJECTION_HIBERNATE;C:\Temp\sql2-custom.txt|SQL_INJECTION_HIBERNATE"</li>
+     *     <li>-Dfindsecbugs.injection.customconfigfile.ScriptInjectionDetector="C:\Temp\script-engine-custom.txt|SCRIPT_ENGINE_INJECTION;C:\Temp\el-custom.txt|EL_INJECTION"</li>
+     * </ul>
+     */
+    protected void loadCustomConfigFiles() {
+        String customConfigFile = FindSecBugsGlobalConfig.loadFromSystem("findsecbugs.injection.customconfigfile." + getClass().getSimpleName(), null);
```
Comment (h3xstream) [inline plugin/src/main/java/com/h3xstream/findsecbugs/injection/BasicInjectionDetector.java]: I would prefer that the specialized configuration would be obtain from a non-static method. `loadFromSystem()` could be private.
The singleton pattern can be quite useful when testing the effect of configuration changes.

For now:
```java
FindSecBugsGlobalConfig.getInstance().getCustomConfigFile( getClass().getSimpleName() )
```
If the configuration gets tested:
```java
protected FindSecBugsGlobalConfig config = FindSecBugsGlobalConfig.getInstance()
[...]
config.getCustomConfigFile( getClass().getSimpleName() );
```

**Comment by h3xstream** [comment]
Looks good !

- The support for cast in numeric value is a really good idea.
- Good Javadoc in general. It is very clear.
- Since you added Portlet api behavior in the taint-config directory, I made some search on it. It gave me some ideas: #216. Or maybe this was already integrated in your changes..
- I like the renaming from TaintMethodSummary to TaintConfig

File: plugin/src/main/java/com/h3xstream/findsecbugs/injection/BasicInjectionDetector.java (hunk 1 above)
Comment (topolik) [inline plugin/src/main/java/com/h3xstream/findsecbugs/injection/BasicInjectionDetector.java]: Please see https://github.com/find-sec-bugs/find-sec-bugs/pull/218

**Comment by topolik** [comment]
Thank you for the review and merge!

I didn't implement XSS for portlet API yet, but it's active from inside JSPs already.

The sinks for XSS detectors are a bit simpler and not suitable for Liferay use cases. The sinks - InputStream/Reader objects from ServletRequest and PortletRequest - can be used outside JSPs, Servlets and Portlets, so current detectors are not able to catch these uses. For example Spring Portlet MVC is driven by annotations, no Servlet no Portlet classes, all we have is PortletRequest from which we can recognize the sink.

Beside, I'd like to include XSS checking inside JSP taglibs, it seems to be missing now.

**Pull request merged**